File: src/report/text.ts
import type { WorktreeRemovalFailure } from "../targets.js";
import {
  displayClass,
  formatDelta,
  formatEvidence,
  formatHintLabel,
  formatLabel,
  formatVariantName,
  formatVerdictDelta,
  GATED_GEOMEAN_LABEL,
  getGlyph,
  hasUnstableHighlight,
  type HighlightBlock,
  highlightLabel,
  pluralize,
  selectHighlights,
  type Style,
  styleWithin,
  truncateLabels,
  UNSTABLE_FUTILITY_NOTE,
  verdictSummaryParts,
  VERDICT_STYLES,
  footerLines,
  withColor,
  withDisplayLabels,
} from "./format.js";
import { spansManyKinds } from "./sections.js";
import { renderMeasureTable } from "./text-measure.js";
import { renderComparisonTable } from "./text-multi.js";
import { renderTable } from "./text-single.js";
import { styleGlyphAndDelta, styleSpans } from "./text-table-core.js";
import type {
  CandidateComparison,
  ComparisonResult,
  FailOnCondition,
  MeasurementResult,
  MetricComparisons,
  ReportOptions,
  WorktreeCleanupOutcome,
} from "./types.js";

// ---------------------------------------------------------------------------
// Header
// ---------------------------------------------------------------------------

/** The `·` separator in the report header, dimmed in colored mode. */
const HEADER_SEPARATOR = "·";

/** Join a run header's parts with the dimmed `·` separator every report header shares. */
function joinHeaderParts(parts: readonly string[]): string {
  return parts.join(` ${formatLabel(HEADER_SEPARATOR, ["dim"])} `);
}

// ---------------------------------------------------------------------------
// Highlights
// ---------------------------------------------------------------------------

/** One line tallying every verdict class one candidate earned. */
function renderSummary(metrics: MetricComparisons, candidateIndex: number): string {
  return verdictSummaryParts(metrics, candidateIndex).join("   ");
}

/** Gap between the longest highlighted metric name and the delta that follows it. */
const HIGHLIGHT_NAME_GUTTER = 2;

/**
 * Width the highlights block right-aligns its deltas in — the length of a
 * `±NN.N%` percentage.
 */
const HIGHLIGHT_DELTA_WIDTH = 6;

/** The heading a candidate's non-empty highlights list opens with. */
const HIGHLIGHTS_HEADING = "highlights";

function highlightEntries(metrics: MetricComparisons, candidateIndex: number): HighlightBlock {
  const highlights = selectHighlights(metrics, candidateIndex);
  if (highlights.length === 0) return { entries: [], unstable: false };

  const qualify = spansManyKinds(metrics);
  const labeled = highlights.map((highlight) => ({
    highlight,
    label: highlightLabel(highlight, qualify),
  }));
  const nameWidth = Math.max(...labeled.map(({ label }) => label.length)) + HIGHLIGHT_NAME_GUTTER;

  const entries = labeled.map(({ highlight: { metric, candidate }, label }) => {
    const verdict = candidate.verdict;
    const shown = displayClass(verdict);
    const delta = formatVerdictDelta(verdict);
    const evidence = formatEvidence(verdict, metric.meta.unit, metric.baselineMedian);
    const suffix = evidence === "" ? "" : `  ${evidence}`;

    const plain = `  ${getGlyph(shown)} ${label.padEnd(nameWidth)}${delta.padStart(
      HIGHLIGHT_DELTA_WIDTH,
    )}${suffix}`;

    const style = VERDICT_STYLES[shown];
    let styled = styleGlyphAndDelta(plain, {
      shown,
      delta,
      style,
      deltaSearch: { last: true },
    });
    if (evidence !== "") {
      styled = styleWithin(styled, evidence, ["dim"]);
    }
    return styled;
  });

  return { entries, unstable: hasUnstableHighlight(highlights) };
}

/** The futility note, indented to sit under the entries it qualifies. */
function futilityLine(): string {
  return `  ${formatLabel(UNSTABLE_FUTILITY_NOTE, ["dim"])}`;
}

/** The glyph flagging a gate the run's own `--fail-on` conditions would trip. */
const GATE_TRIP_GLYPH = "⚑";

function gateTripLines(
  candidate: CandidateComparison,
  conditions: readonly FailOnCondition[],
): string[] {
  const thresholds = conditions.flatMap((condition) =>
    condition.kind === "geomean" ? [condition.pct] : [],
  );

  return candidate.kinds.flatMap((kind) => {
    const geomean = kind.gatedGeomean;
    if (geomean === undefined || geomean.n === 0) return [];

    const delta = formatDelta(geomean.value);
    const style: Style = VERDICT_STYLES.regressed;
    return thresholds
      .filter((pct) => geomean.value >= pct)
      .map((pct) => {
        const plain = `  ${GATE_TRIP_GLYPH} ${kind.kind} ${GATED_GEOMEAN_LABEL} ${delta} exceeded --fail-on geomean:${pct}`;
        return styleSpans(plain, [
          { text: GATE_TRIP_GLYPH, style },
          { text: delta, style },
        ]);
      });
  });
}

function highlightSection(blocks: readonly HighlightBlock[]): string[] {
  const nonEmpty = blocks.filter((block) => block.entries.length > 0);
  if (nonEmpty.length === 0) return [];

  const lines = [formatLabel(HIGHLIGHTS_HEADING, ["bold"])];
  for (const block of nonEmpty) {
    if (block.label === undefined) {
      lines.push(...block.entries);
    } else {
      lines.push(
        `  ${formatLabel(block.label, ["bold"])}`,
        ...block.entries.map((entry) => `  ${entry}`),
      );
    }
  }
  if (nonEmpty.some((block) => block.unstable)) lines.push(futilityLine());
  return lines;
}

function renderHighlights(
  metrics: MetricComparisons,
  candidateIndex: number,
  gateTrips: readonly string[],
): string[] {
  const { entries, unstable } = highlightEntries(metrics, candidateIndex);
  return highlightSection([{ entries: [...entries, ...gateTrips], unstable }]);
}

function renderCandidateHighlights(
  result: ComparisonResult,
  conditions: readonly FailOnCondition[],
): string[] {
  const blocks = result.candidates.map((candidate, index) => {
    const { entries, unstable } = highlightEntries(result.metrics, index);
    return {
      label: candidate.label,
      entries: [...entries, ...gateTripLines(candidate, conditions)],
      unstable,
    };
  });
  return highlightSection(blocks);
}

function renderSummaries(result: ComparisonResult): string[] {
  const labelWidth = Math.max(...result.candidates.map((candidate) => candidate.label.length));
  return result.candidates.map((candidate, index) => {
    const paddedLabel = candidate.label.padEnd(labelWidth);
    const styledLabel = formatLabel(paddedLabel, ["bold"]);
    return `${styledLabel}  ${renderSummary(result.metrics, index)}`;
  });
}

// ---------------------------------------------------------------------------
// Footers
// ---------------------------------------------------------------------------

function renderComparison(
  result: ComparisonResult,
  conditions: readonly FailOnCondition[],
): string[] {
  const lines = [...renderComparisonTable(result), "", ...renderSummaries(result)];

  const highlights = renderCandidateHighlights(result, conditions);
  if (highlights.length > 0) {
    lines.push("", ...highlights);
  }

  return lines;
}

function renderMethodFooter(result: ComparisonResult, verbose: boolean): string[] {
  return footerLines(result.metrics, verbose, (hint) => `${formatHintLabel()} ${hint}`);
}

/**
 * Collapse a git diagnostic onto one line.
 *
 * git routinely emits several lines for one failure — a `warning:` line before the
 * `fatal:` line, plus indented continuations.
 */
function toSingleLine(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}

/** Format worktree removal failures and prune errors into indented diagnostic lines. */
export function formatCleanupFailures(
  leftBehind: readonly WorktreeRemovalFailure[],
  pruneError: string | undefined,
): string[] {
  const lines = leftBehind.map(
    (failure) => `  left behind: ${failure.dir} (${toSingleLine(failure.error)})`,
  );

  if (pruneError !== undefined) {
    lines.push(`  worktree prune failed: ${toSingleLine(pruneError)}`);
  }

  return lines;
}

function renderWorktreeFooter(result: WorktreeCleanupOutcome): string[] {
  const details = formatCleanupFailures(result.worktreesLeftBehind, result.worktreePruneError);
  if (details.length === 0) return [];

  return [
    `${pluralize(result.worktreesRemoved, "worktree")} removed · ${result.worktreesLeftBehind.length} left behind`,
    ...details,
  ];
}

function renderCandidate(
  result: ComparisonResult,
  candidate: CandidateComparison,
  candidateIndex: number,
  conditions: readonly FailOnCondition[],
): string[] {
  const lines = [
    ...renderTable(result, candidate, candidateIndex),
    "",
    renderSummary(result.metrics, candidateIndex),
  ];

  const highlights = renderHighlights(
    result.metrics,
    candidateIndex,
    gateTripLines(candidate, conditions),
  );
  if (highlights.length > 0) {
    lines.push("", ...highlights);
  }

  return lines;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Pluralize the "paired sample" label used in comparison report headers. */
export function pairedSamples(samples: number): string {
  return pluralize(samples, "paired sample");
}

function compareHeader(display: ComparisonResult): string {
  const candidateNames = display.candidates
    .map((candidate) => formatVariantName(candidate.label))
    .join(", ");
  return joinHeaderParts([
    formatLabel("gymrat compare", ["bold"]),
    `baseline ${formatVariantName(display.baselineLabel)} ↔ ${candidateNames}`,
    pairedSamples(display.samples),
    `adapter: ${display.adapter}`,
  ]);
}

export function renderReport(result: ComparisonResult, options: ReportOptions = {}): string {
  return withColor(options.color, () => {
    const display = withDisplayLabels(result);
    const lines = [options.header ?? compareHeader(display)];

    const conditions = options.failOn ?? [];
    if (display.candidates.length > 1) {
      lines.push(...renderComparison(display, conditions));
    } else {
      const candidate = display.candidates[0];
      if (candidate === undefined) throw new Error("expected at least one candidate");
      lines.push(...renderCandidate(display, candidate, 0, conditions));
    }

    const footer = [
      ...renderMethodFooter(display, options.verbose ?? false),
      ...renderWorktreeFooter(display),
    ];
    if (footer.length > 0) {
      lines.push("", ...footer);
    }

    return lines.join("\n");
  });
}

export function renderMeasureReport(
  result: MeasurementResult,
  options: ReportOptions = {},
): string {
  return withColor(options.color, () => {
    const label = truncateLabels([result.label])[0] ?? result.label;
    const header = joinHeaderParts([
      formatLabel("gymrat measure", ["bold"]),
      formatVariantName(label),
      pluralize(result.samples, "sample"),
      `adapter: ${result.adapter}`,
    ]);

    const lines = [header, ...renderMeasureTable(result, label)];

    const footer = renderWorktreeFooter(result);
    if (footer.length > 0) {
      lines.push("", ...footer);
    }

    return lines.join("\n");
  });
}

File: src/gymrat_py/report/text.py
"""The human-readable text reports: the compare report and the measure report.

This module owns only the run header and the table dispatch: one candidate draws
the single-candidate table, two or more the multi-candidate table. The verdict
summary, highlights, gate trips and footers that sit around a full comparison
report are added by a later task; a comparison report here is the header and the
comparison table alone, and a measure report the header and the measurement
table.
"""

from __future__ import annotations

from dataclasses import replace
from typing import TYPE_CHECKING

from rich.markup import escape
from rich.text import Text

from gymrat_py.report.format import pluralize
from gymrat_py.report.style import VARIANT_NAME_STYLE, render_lines, truncate_labels
from gymrat_py.report.table import markup
from gymrat_py.report.text_measure import render_measure_table
from gymrat_py.report.text_multi import render_comparison_table
from gymrat_py.report.text_single import render_table
from gymrat_py.report.types import ReportOptions

if TYPE_CHECKING:
    from gymrat_py.report.types import ComparisonResult, MeasurementResult

# A wide render width so the header line never soft-wraps.
_RENDER_WIDTH = 200

# The default presentation flags: detect color, no header override. Immutable, so
# one shared instance is safe as a default argument.
_DEFAULT_OPTIONS = ReportOptions()

# The `·` separator every report header joins its parts with, dimmed in color.
_HEADER_SEPARATOR = "·"


def _join_header_parts(parts: list[str]) -> str:
    """Join header parts with the dimmed ``·`` separator every report header shares."""
    return f" {markup(_HEADER_SEPARATOR, 'dim')} ".join(parts)


def _render_line(text: str, *, color: bool | None) -> str:
    """Resolve a markup line to text once, deferring wrapping so the header stays whole."""
    return render_lines(Text.from_markup(text), color=color, width=_RENDER_WIDTH)


def with_display_labels(result: ComparisonResult) -> ComparisonResult:
    """``result`` with every variant label replaced by the name the report prints.

    The baseline and candidate labels are shortened together, so a label prints
    the same way wherever the report names it — the header, the column it heads,
    the geomean row. Metric names are left whole; only the variant labels shorten.

    Args:
        result: The comparison to relabel.

    Returns:
        A copy with shortened variant labels.
    """
    labels = truncate_labels([result.baseline_label, *(c.label for c in result.candidates)])
    return replace(
        result,
        baseline_label=labels[0],
        candidates=tuple(
            replace(candidate, label=labels[index + 1])
            for index, candidate in enumerate(result.candidates)
        ),
    )


def paired_samples(samples: int) -> str:
    """The ``N paired samples`` label the comparison report header carries."""
    return pluralize(samples, "paired sample")


def _compare_header(display: ComparisonResult) -> str:
    """The compare report's run header as markup: the baseline's role, the variants, the run."""
    candidate_names = ", ".join(
        markup(candidate.label, VARIANT_NAME_STYLE) for candidate in display.candidates
    )
    return _join_header_parts(
        [
            markup("gymrat compare", "bold"),
            f"baseline {markup(display.baseline_label, VARIANT_NAME_STYLE)} ↔ {candidate_names}",
            escape(paired_samples(display.samples)),
            f"adapter: {escape(display.adapter)}",
        ]
    )


def render_report(result: ComparisonResult, options: ReportOptions = _DEFAULT_OPTIONS) -> str:
    """Render a comparison report: the run header and the comparison table.

    One candidate draws the single-candidate table; two or more draw the
    multi-candidate table. The verdict summary, highlights, gate trips and footers
    arrive with a later task. ``options.header`` replaces the run header verbatim;
    ``options.color`` forces color on or off, or defers to the environment when
    ``None``.

    Args:
        result: The comparison to draw.
        options: The presentation flags.

    Returns:
        The rendered report.
    """
    color = options.color
    display = with_display_labels(result)

    if options.header is not None:
        lines = [options.header]
    else:
        lines = [_render_line(_compare_header(display), color=color)]

    if len(display.candidates) == 1:
        lines.extend(render_table(display, display.candidates[0], 0, color=color))
    elif len(display.candidates) > 1:
        lines.extend(render_comparison_table(display, color=color))

    return "\n".join(lines)


def render_measure_report(
    result: MeasurementResult,
    options: ReportOptions = _DEFAULT_OPTIONS,
) -> str:
    """Render a single-target measurement report: the run header and the measurement table.

    Args:
        result: The measurement to draw.
        options: The presentation flags. ``options.color`` forces color on or off,
            or defers to the environment when ``None``.

    Returns:
        The rendered report.
    """
    color = options.color
    label = truncate_labels([result.label])[0]
    header = _join_header_parts(
        [
            markup("gymrat measure", "bold"),
            markup(label, VARIANT_NAME_STYLE),
            escape(pluralize(result.samples, "sample")),
            f"adapter: {escape(result.adapter)}",
        ]
    )

    lines = [_render_line(header, color=color)]
    lines.extend(render_measure_table(result, label, color=color))
    return "\n".join(lines)
